In [1]:
import trainer.glue_base as glue_base
import models.sparse_token as sparse
import pickle, importlib
importlib.reload(glue_base)
importlib.reload(sparse)
Glue = glue_base.GlueAttentionApproxTrainer
PICKLE_PATH = "glue_benchmark_wiki.pkl"
TEX_PATH = "glue_benchmark_wiki.tex"

In [2]:
subsets = ["cola","mnli","mrpc","qnli","qqp","rte","sst2","stsb","wnli",]
kss = [
    0.1, 0.25, 0.375, 0.5, 0.625, 0.75, 0.875, 0.999, 'dynamic', 
    'dynamic:avg:avg:true', 'dynamic:avg:avg:false', 'dynamic:avg:max:true', 'dynamic:avg:max:false',
    'dynamic:max:avg:true', 'dynamic:max:avg:false', 'dynamic:max:max:true', 'dynamic:max:max:false',
]
kss = ['dynamic']
sparse.benchmark_reset()
# subsets = ["mrpc"]
# kss = ['dynamic:avg:avg:f',0.1]

def get_score(score):
    if 'accuracy' in score:
        return score['accuracy'], "acc"
    first_metric = list(score.keys())[0]
    return score[first_metric], first_metric

results = {}
i = 0
for subset in subsets:
    trainer = Glue(dataset=subset, factor=16, batch_size=-1, device=0, wiki_train=True)
    trainer.load()
    scores = {}
    metric_name = ""
    bert_score, metric_name = get_score(trainer.eval_base_model())
    scores['bert'] = f'{bert_score:.5f}'
    for ks in kss:
        sparse.benchmark_reset()
        sparse_score, _ = get_score(trainer.eval_sparse_model(ks=ks))
        if isinstance(ks, str) and ks.startswith('dynamic'):
            est_k = sparse.benchmark_get_average('est_k')
            scores[str(ks)] = f'{sparse_score:.5f} (k:{est_k:.2f})'
        else:
            scores[str(ks)] = f'{sparse_score:.5f}'
        i += 1
        count = len(subsets) * len(kss)
        print(f'{i}/{count}')
    results[f"{subset} ({metric_name})"] = scores

with open(PICKLE_PATH, 'wb') as f:
    pickle.dump(results, f)

sparse.benchmark_report()

Trainer: cola


Reusing dataset glue (/home/ainl/.cache/huggingface/datasets/glue/cola/1.0.0/dacbe3125aa31d7f70367a07a8a9e72a5a0bfeb5fc42e75c9db75b96da6053ad)
Loading cached processed dataset at /home/ainl/.cache/huggingface/datasets/glue/cola/1.0.0/dacbe3125aa31d7f70367a07a8a9e72a5a0bfeb5fc42e75c9db75b96da6053ad/cache-f0e04bff3297c11c.arrow
Loading cached processed dataset at /home/ainl/.cache/huggingface/datasets/glue/cola/1.0.0/dacbe3125aa31d7f70367a07a8a9e72a5a0bfeb5fc42e75c9db75b96da6053ad/cache-b9e7d91153cc59cf.arrow
Reusing dataset glue (/home/ainl/.cache/huggingface/datasets/glue/cola/1.0.0/dacbe3125aa31d7f70367a07a8a9e72a5a0bfeb5fc42e75c9db75b96da6053ad)
Loading cached processed dataset at /home/ainl/.cache/huggingface/datasets/glue/cola/1.0.0/dacbe3125aa31d7f70367a07a8a9e72a5a0bfeb5fc42e75c9db75b96da6053ad/cache-7e570c95fb4f22e4.arrow
Loading cached processed dataset at /home/ainl/.cache/huggingface/datasets/glue/cola/1.0.0/dacbe3125aa31d7f70367a07a8a9e72a5a0bfeb5fc42e75c9db75b96da6053ad/cac

Trainer: Save checkpoint path saves/glue-cola-16-wiki.pth


eval: 100%|██████████| 131/131 [00:01<00:00, 72.15it/s]


metric score {'matthews_correlation': 0.5338774230813111}
avg occupy 0.34522852946633237


eval: 100%|██████████| 131/131 [00:03<00:00, 37.14it/s]


metric score {'matthews_correlation': 0.5338774230813111}
avg occupy 0.34522852946633237
1/9
Trainer: mnli


100%|██████████| 10/10 [00:00<00:00, 15.11ba/s]


Trainer: Save checkpoint path saves/glue-mnli-16-wiki.pth


eval: 100%|██████████| 1227/1227 [00:28<00:00, 42.31it/s]


metric score {'accuracy': 0.8419765664798777}
avg occupy 0.2057366697815377


eval: 100%|██████████| 1227/1227 [00:36<00:00, 34.07it/s]


metric score {'accuracy': 0.8155883851248089}
avg occupy 0.2057366697815377
2/9
Trainer: mrpc


Trainer: Save checkpoint path saves/glue-mrpc-16-wiki.pth


eval: 100%|██████████| 216/216 [00:03<00:00, 71.15it/s]


metric score {'accuracy': 0.8440579710144928, 'f1': 0.8865457612821593}
avg occupy 0.5399536083129833


eval: 100%|██████████| 216/216 [00:05<00:00, 38.69it/s]


metric score {'accuracy': 0.7466666666666667, 'f1': 0.7941592086669808}
avg occupy 0.5399536083129833
3/9
Trainer: qnli


Trainer: Save checkpoint path saves/glue-qnli-16-wiki.pth


eval: 100%|██████████| 683/683 [00:13<00:00, 49.12it/s]


metric score {'accuracy': 0.9154310818231741}
avg occupy 0.30243502260079336


eval: 100%|██████████| 683/683 [00:19<00:00, 34.30it/s]


metric score {'accuracy': 0.9044481054365733}
avg occupy 0.30243502260079336
4/9
Trainer: qqp


100%|██████████| 41/41 [00:01<00:00, 20.74ba/s]


Trainer: Save checkpoint path saves/glue-qqp-16-wiki.pth


eval: 100%|██████████| 5054/5054 [01:18<00:00, 64.61it/s]


metric score {'accuracy': 0.9090774177590898, 'f1': 0.8781732617485252}
avg occupy 0.2734741473281857


eval: 100%|██████████| 5054/5054 [02:15<00:00, 37.26it/s]


metric score {'accuracy': 0.9089290131090774, 'f1': 0.8781440296531638}
avg occupy 0.2734741473281857
5/9
Trainer: rte


100%|██████████| 1/1 [00:00<00:00, 38.15ba/s]


Trainer: Save checkpoint path saves/glue-rte-16-wiki.pth


eval: 100%|██████████| 35/35 [00:01<00:00, 32.27it/s]


metric score {'accuracy': 0.7256317689530686}
avg occupy 0.26989271615199084


eval: 100%|██████████| 35/35 [00:01<00:00, 28.07it/s]


metric score {'accuracy': 0.7075812274368231}
avg occupy 0.26989271615199084
6/9
Trainer: sst2


100%|██████████| 1/1 [00:00<00:00, 38.95ba/s]


Trainer: Save checkpoint path saves/glue-sst2-16-wiki.pth


eval: 100%|██████████| 109/109 [00:01<00:00, 95.94it/s]


metric score {'accuracy': 0.9243119266055045}
avg occupy 0.4575271059216015


eval: 100%|██████████| 109/109 [00:02<00:00, 38.31it/s]


metric score {'accuracy': 0.9220183486238532}
avg occupy 0.4575271059216015
7/9
Trainer: stsb


100%|██████████| 2/2 [00:00<00:00, 31.44ba/s]


Trainer: Save checkpoint path saves/glue-stsb-16-wiki.pth


eval: 100%|██████████| 188/188 [00:02<00:00, 82.26it/s]


metric score {'pearson': 0.8804685139192188, 'spearmanr': 0.8763316039868316}
avg occupy 0.387254677182685


eval: 100%|██████████| 188/188 [00:05<00:00, 36.91it/s]


metric score {'pearson': 0.8657197450015853, 'spearmanr': 0.8597012953718358}
avg occupy 0.387254677182685
8/9
Trainer: wnli


100%|██████████| 1/1 [00:00<00:00, 172.97ba/s]


Trainer: Save checkpoint path saves/glue-wnli-16-wiki.pth


eval: 100%|██████████| 9/9 [00:00<00:00, 58.27it/s]


metric score {'accuracy': 0.5633802816901409}
avg occupy 0.3578609244533317


eval: 100%|██████████| 9/9 [00:00<00:00, 36.04it/s]

metric score {'accuracy': 0.5633802816901409}
avg occupy 0.3578609244533317
9/9
est_k: (108, 58.038095238095245), avg: 0.5373897707231041


In [14]:
import pickle
import pandas as pd

with open(PICKLE_PATH, 'rb') as f:
    results = pickle.load(f)
with open('glue_benchmark.pkl', 'rb') as f:
    results_original = pickle.load(f)

def convert_data_to_results(results):
    data = []
    subsets = list(results.keys())
    factors = list(results[subsets[0]].keys())
    for factor in factors:             
        row = []
        for subset in subsets:
            row.append(results[subset][factor])
        data.append(row)
    return data, factors

data, factors = convert_data_to_results(results)
data_origin, _ = convert_data_to_results(results_original)

df = pd.DataFrame(data + [data_origin[9]], columns=subsets, index=factors+["dynamic:original"])
tex = df.to_latex()
with open(TEX_PATH, 'w') as f:
    f.write(tex)
df

/tmp/ipykernel_89525/225414492.py:24: FutureWarning: In future versions `DataFrame.to_latex` is expected to utilise the base implementation of `Styler.to_latex` for formatting and rendering. The arguments signature may therefore change. It is recommended instead to use `DataFrame.style.to_latex` which also contains additional functionality.
  tex = df.to_latex()


cola (matthews_correlation)        mnli (acc)  \
bert                                 0.53388           0.84198   
dynamic                     0.53388 (k:0.48)  0.81559 (k:0.30)   
dynamic:original            0.53388 (k:0.48)  0.84004 (k:0.33)   

                        mrpc (acc)        qnli (acc)         qqp (acc)  \
bert                       0.84406           0.91543           0.90908   
dynamic           0.74667 (k:0.66)  0.90445 (k:0.41)  0.90893 (k:0.43)   
dynamic:original  0.83942 (k:0.71)  0.91287 (k:0.44)  0.90920 (k:0.45)   

                         rte (acc)        sst2 (acc)    stsb (pearson)  \
bert                       0.72563           0.92431           0.88047   
dynamic           0.70758 (k:0.38)  0.92202 (k:0.69)  0.86572 (k:0.49)   
dynamic:original  0.72563 (k:0.55)  0.92317 (k:0.73)  0.88043 (k:0.51)   

                        wnli (acc)  
bert                       0.56338  
dynamic           0.56338 (k:0.54)  
dynamic:original  0.56338 (k:0.63)

# pre data augment

cola (matthews_correlation)	mnli (acc)	mrpc (acc)	qnli (acc)	qqp (acc)	rte (acc)	sst2 (acc)	stsb (pearson)	wnli (acc)
bert	0.53388	0.84198	0.84406	0.91543	0.90908	0.72563	0.92431	0.88047	0.56338
dynamic	0.53388 (k:0.48)	0.81559 (k:0.30)	0.74667 (k:0.66)	0.90445 (k:0.41)	0.90893 (k:0.43)	0.70758 (k:0.38)	0.92202 (k:0.69)	0.86572 (k:0.49)	0.56338 (k:0.54)
dynamic:original	0.53388 (k:0.48)	0.84004 (k:0.33)	0.83942 (k:0.71)	0.91287 (k:0.44)	0.90920 (k:0.45)	0.72563 (k:0.55)	0.92317 (k:0.73)	0.88043 (k:0.51)	0.56338 (k:0.63)